lib/gears/surface.lua

```diff
@@ -283,6 +284,79 @@ function surface.widget_to_surface(widget, width, height, format)
     return img, run_in_hierarchy(widget, cr, width, height)
 end
 
+--- Crop a surface on its edges
+-- @tparam[opt=nil] table args
+-- @tparam int args.left Left cutoff
+-- @tparam int args.right Right cutoff
+-- @tparam int args.top Top cutoff
+-- @tparam int args.bottom Bottom cutoff
+-- @tparam[opt=nil] surface args.surf the surface to crop
+-- @return the cropped surface
+-- @staticfct crop_surface
+function surface.crop_surface(args)
+    args = args or {}
+    if not args.surface then
+        gdebug.print_error("No surface to crop_surface supplied")
+        return nil
+    end
+    local surf = args.surface
+    local left = args.left or 0
+    local right = args.right or 0
+    local top = args.top or 0
+    local bottom = args.bottom or 0
+
+    local old_w, old_h = surface.get_size(surf)
+    local new_w, new_h = old_w - left - right, old_h - top - bottom
+    -- breaking stuff with cairo crashes awesome with no way to restart in place
+    -- so here are checks for user error
+    if new_w < 0 or new_h < 0 then
+        gdebug.print_error("crop_surface target size too small")
+        return nil
+    end
+
+    local ret = cairo.ImageSurface(cairo.Format.ARGB32, new_w, new_h)
+    local cr = cairo.Context(ret)
+    cr:set_source_surface(surf, -left, -top)
+    cr.operator = cairo.Operator.SOURCE
+    cr:paint()
+
+    return ret
+end
+
+--- Crop a surface to a given ratio
+-- This creates a copy of the surface, as surfaces seem to not be
+-- resizable after creation, so you have to use the functions return
+-- value. The resulting surface is in the center of the original surface
+-- @param surf surface the Cairo surface to crop
+-- @param ratio number the ratio the image should be cropped to (widht/height)
+-- @return the a cropped copy of the input surface
+-- @staticfct crop_surface_by_ratio
+function surface.crop_surface_by_ratio(surf, ratio)
```

Comment:
now this became weird.

just leave `surface.crop_surface` with `ratio` as an argument

Reply:
how should something like
```
surface.crop_surface({
    surface = surface.
    ratio = 0.5,
    left = 5
})
```
be handled then? Should the image be cropped before the "ratio + zoomed" operation that the use of ratio implies is performed?

My thoughts:
The surface by ratio crop makes sense from a users perspective if the goal is to crop an image for use as wallpaper for example: You just pass in your screen ratio and you have a wallpaper that fits your screen.
The more manual approach could make more sense for someone that maybe wants to crop an image with a fixed size (buttons(?)) to fit a widget size exactly.

Mixing both crop options could confuse a lot of users and you need to explain too much about how this works in the back in the docs to prevent people explaining about unwanted or random behavior.

Reply:
it's more about shared implementation - switching between pixel and ratio modes basically changes only one two line of the whole function

so answering your question, how to handle that particular case - either apply both ratio and pixel crop after computing the ratio or just raise error that args are incompatible

Reply:
btw you don't really need to ratio-crop in case of wallpaper - that could be done on the level of wallpaper module itself (ie underlying `imagebox`)